### PR TITLE
enable bsz>1 with padding

### PR DESCRIPTION
**Pull request by felipemello1**
TODO: fix wrong truncation

image

### CODE REVIEW

File: scripts/runnable_recipe_ray_vllm_weight_sync.py [hunk 1]
```diff
@@ -250,6 +250,12 @@ def setup(self, cfg):
             actor_options={"num_cpus": 10, "num_gpus": 0},
             maxsize=self.cfg.vllm.queue_maxsize,
         )
+
+        # avoid sampling with replacement
```
Comment (Darktex) [inline scripts/runnable_recipe_ray_vllm_weight_sync.py:254]: We don't want to avoid it all the time though. Maybe `log.info`? It's perfectly fine and expected to sample with replacement

File: torchtune/dev/grpo/types.py [hunk 2]
```diff
@@ -5,9 +5,203 @@
...
+    Represents a collection of unscored trajectories generated from the same base prompt.
+
+    Attributes:
+        trajectories (List[UnscoredTrajectory]): A list of individual unscored trajectories
+            belonging to this group.
+        group_id (str): The unique identifier shared by all trajectories in this group.
+        total_tokens (int): The sum of prompt lengths and response lengths for all
+            trajectories in the group.
+    """
+    trajectories: List[UnscoredTrajectory]
```
Comment (Darktex) [inline torchtune/dev/grpo/types.py:54]: Yeah, let's not nest. It's just too annoying to deal with. Inheriting dataclasses is also a bit icky, but at least it's easier to work with as fields aren't nested

File: torchtune/dev/grpo/types.py [hunk 3]
```diff
@@ -5,9 +5,203 @@
...
+            successes=self.successes.to(device),
+            answer=self.answer,
+            sequence_id=self.sequence_id,
+            group_id=self.group_id,
+            reward_metadata=self.reward_metadata,
+            policy_version=self.policy_version
+        )
+
+@dataclass
+class PackedTrajectory:
```
Comment (Darktex) [inline torchtune/dev/grpo/types.py:124]: Probably worth inheriting here

File: torchtune/dev/grpo/types.py [hunk 4]
```diff
@@ -5,9 +5,203 @@
...
+    packed_seq_len: int
+    ref_logprobs: Optional[torch.Tensor] = None
+    advantages: Optional[torch.Tensor] = None
+    targets: Optional[torch.Tensor] = None
+    sequence_ids: Optional[List[str]] = None
+    group_ids: Optional[List[str]] = None
+    actual_total_tokens: int = 0
+
+@dataclass
+class GRPOStats:
```
Comment (Darktex) [inline torchtune/dev/grpo/types.py:178]: Can be made more generic: there's nothing inherently about GRPO as an algo compared to eg PPO or any other online RL algo. Maybe `TrainingStats`?

File: torchtune/dev/grpo/types.py [hunk 5]
```diff
@@ -5,9 +5,203 @@
...
+    loss: torch.Tensor
+    policy_loss: torch.Tensor
+    kl_loss: torch.Tensor
+    ratios: torch.Tensor
+    clipfrac: torch.Tensor
+    approx_policy_kls: torch.Tensor
+    metadata: Optional[Dict[str, Any]] = None
 
 
 class GRPOTrajectory(NamedTuple):
```
Comment (Darktex) [inline torchtune/dev/grpo/types.py:207]: Can be nuked? Just has one field

File: torchtune/dev/rl/datatypes/trajectory.py [hunk 6]
```diff
@@ -1,20 +1,152 @@
 from typing import Dict, List
 
+from dataclasses import dataclass
 import torch
+from typing import List, Dict, Optional, Any
+
 from tensordict import TensorClass
 
+@dataclass
+class UnscoredTrajectory:
```
Comment (Darktex) [inline torchtune/dev/rl/datatypes/trajectory.py:10]: Copypasta? It should exist only in one place :D

File: torchtune/dev/rl/datatypes/trajectory.py [hunk 7]
```diff
@@ -1,20 +1,152 @@
...
-    reward_metadata: Dict[str, List[str]]
-    sequence_ids: List[str]
+    answer: str
+    sequence_id: str
+    group_id: str
+    reward_metadata: Dict[str, Any]
+    policy_version: int
+
+@dataclass
+class PackedTrajectory:
```
Comment (Darktex) [inline torchtune/dev/rl/datatypes/trajectory.py:101]: Copypasta? It should exist only in one place :D

File: torchtune/dev/rl/workers/datacollectors/sync.py [hunk 8]
```diff
@@ -134,60 +135,92 @@ def remote_weight_updater(self, value: RemoteWeightUpdaterBase | None):
...
-        answers = data["answers"]
-        if hasattr(
-            self.inference_server.llm_engine.model_executor.driver_worker.worker,
-            "policy_version",
-        ):
-            policy_version = (
-                self.inference_server.llm_engine.model_executor.driver_worker.worker.policy_version.item()
+        
+        # Data extraction and unpadding
+        prompt_tokens_padded = data['tokens'].cpu()              # [B, longest_prompt]
```
Comment (Darktex) [inline torchtune/dev/rl/workers/datacollectors/sync.py:166]: Thank you for adding size comments ❤️

File: torchtune/dev/rl/workers/datacollectors/sync.py [hunk 9]
```diff
@@ -134,60 +135,92 @@ def remote_weight_updater(self, value: RemoteWeightUpdaterBase | None):
...
-            successes=None,
-            reward_metadata=None,
-            sequence_ids=sequence_ids,
-        )
+            # group trajectories by group_id
+            trajectories[traj.group_id].append(traj)
+
+        
+        # create GroupedUnscoredTrajectories for each group
+        grouped_trajectories = []
```
Comment (Darktex) [inline torchtune/dev/rl/workers/datacollectors/sync.py:212]: Eventually this will need to be enabled if Monte Carlo-based advantage estimation is turned on, otherwise it will be off. EG in "normal" PPO it's off

**Review by Darktex — CHANGES_REQUESTED** [review]
A few copypastas to remove, but it's not clear to me yet how we go around working with the ReplayBuffer. These dataclasses can give us a way out of the need to pad, but I keep wondering if better support for these datatypes in TensorDict might help us more. I do think that TensorDicts make your life a lot easier...
